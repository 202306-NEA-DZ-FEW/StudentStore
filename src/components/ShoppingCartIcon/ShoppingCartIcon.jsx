import Link from "next/link";
import { FiShoppingCart } from "react-icons/fi";

const ShoppingCartIcon = ({ count }) => {
    return (
        <div>
            <Link className='relative' href='/cart'>
                <FiShoppingCart className="className=' text-[28px] text-[#8685a7] hover:text-orange-500 " />
                {count && (
                    <span className='absolute text-[11px] flex justify-center items-center  font-bold text-white bottom-[20px] w-5 h-5 rounded-full bg-red-500 left-5'>
                        {count}
                    </span>
                )}
            </Link>
        </div>
    );
};

export default ShoppingCartIcon;
