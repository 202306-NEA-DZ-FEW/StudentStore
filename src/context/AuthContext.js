import { auth } from "@/util/firebase";
import {
    onAuthStateChanged,
    signInWithEmailAndPassword,
    signOut,
    sendPasswordResetEmail,
} from "firebase/auth";
import { createContext, useContext, useEffect, useState } from "react";

const AuthContext = createContext();

export function useAuth() {
    return useContext(AuthContext);
}

export function AuthProvider({ children }) {
    const [currentUser, setCurrentUser] = useState();
    const [loading, setLoading] = useState(true);
    // sign in function
    function login(email, password) {
        return signInWithEmailAndPassword(auth, email, password);
    }
    // log out function
    function logout() {
        return signOut(auth);
    }
    // reset email function
    function resetPassword(email) {
        return sendPasswordResetEmail(auth, email);
    }
    // user State function
    useEffect(() => {
        const unsubscribe = onAuthStateChanged(auth, (user) => {
            setCurrentUser(user);
            setLoading(false);
        });
        return unsubscribe;
    }, []);
    const value = {
        currentUser,
        login,
        logout,
        resetPassword,
    };

    return (
        <AuthContext.Provider value={value}>
            {!loading && children}
        </AuthContext.Provider>
    );
}
