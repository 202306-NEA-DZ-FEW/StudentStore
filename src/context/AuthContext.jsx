import { auth } from "@/util/firebase";
import { onAuthStateChanged, signInWithEmailAndPassword } from "firebase/auth";
import { createContext, useContext, useEffect, useState } from "react";

const AuthContext = createContext();

export function useAuth() {
    return useContext(AuthContext);
}

export function AuthProvider({ children }) {
    const [currentUser, setCurrentUser] = useState();
    // sign in function
    function signin(email, password) {
        return signInWithEmailAndPassword(auth, email, password);
    }
    // user State function
    useEffect(() => {
        const unsubscribe = onAuthStateChanged(auth, (user) => {
            setCurrentUser(user);
        });
        return unsubscribe;
    }, []);
    const value = {
        currentUser,
        signin,
    };

    return (
        <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
    );
}
