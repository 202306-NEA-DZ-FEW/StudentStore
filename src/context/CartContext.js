import {
    addDoc,
    collection,
    deleteDoc,
    getDocs,
    onSnapshot,
    query,
    updateDoc,
    where,
} from "firebase/firestore";
import { useTranslation } from "next-i18next";
import { createContext, useEffect, useState } from "react";
import { toast, ToastContainer } from "react-toastify";

import "react-toastify/dist/ReactToastify.css";

import { db } from "@/util/firebase";

import { useAuth } from "./AuthContext";

export const CartContext = createContext({
    cartItems: [],
    addItemToCart: () => {},
    removeItemFromCart: () => {},
    updateCartItem: () => {},
    cartCount: 0,
});

export const CartProvider = ({ children }) => {
    const { t } = useTranslation("common");
    const [cartItems, setCartItems] = useState([]);
    const [cartCount, setCartCount] = useState(0);
    const { currentUser } = useAuth();
    const currentUserUid = currentUser?.uid || "";
    const userId = currentUserUid;
    const cartRef = collection(db, "cart");

    useEffect(() => {
        const newCartCount = cartItems.reduce(
            (total, cartItem) => total + cartItem.quantity,
            0
        );
        setCartCount(newCartCount);
    }, [cartItems]);

    const addItemToCart = async (productToAdd) => {
        if (!currentUser) {
            toast.error(
                t(
                    "You are not logged in. Please log in to add items to your cart."
                )
            );
            return;
        }

        const q = query(
            cartRef,
            where("userId", "==", userId),
            where("productId", "==", productToAdd.id)
        );
        const snapshot = await getDocs(q);

        if (!snapshot.empty) {
            toast.info(t("Item is already in your cart."), {
                autoClose: 2000,
                position: "bottom-center",
            });
            const cartItemDoc = snapshot.docs[0].ref;
            const quantity = snapshot.docs[0].data().quantity;
            await updateDoc(cartItemDoc, { quantity: quantity });
        } else {
            toast.success(t("Item added to your cart."), {
                autoClose: 2000,
                position: "bottom-center",
            });
            await addDoc(cartRef, {
                userId,
                productId: productToAdd.id,
                quantity: 1,
                title: productToAdd.title,
                price: productToAdd.price,
                type: productToAdd.type,
                image: productToAdd.pictures[1],
                category: productToAdd.category,
            });
        }
    };

    const updateCartItem = async (productId, updatedDetails) => {
        const q = query(
            cartRef,
            where("userId", "==", userId),
            where("productId", "==", productId)
        );
        const snapshot = await getDocs(q);

        if (!snapshot.empty) {
            const cartItemDoc = snapshot.docs[0].ref;
            await updateDoc(cartItemDoc, updatedDetails);
        } else {
            toast.error(t("Item not found in your cart."));
        }
    };

    const removeItemFromCart = async (productToDelete) => {
        const q = query(
            cartRef,
            where("userId", "==", userId),
            where("productId", "==", productToDelete)
        );
        const snapshot = await getDocs(q);

        if (!snapshot.empty) {
            toast.success(t("Item removed from your cart."), {
                autoClose: 2000,
            });
            const cartItemDoc = snapshot.docs[0].ref;
            await deleteDoc(cartItemDoc);
        } else {
            toast.error(t("Item not found in your cart."));
        }
    };

    useEffect(() => {
        const fetchCartItems = () => {
            const q = query(cartRef, where("userId", "==", userId));

            const unsubscribe = onSnapshot(q, (snapshot) => {
                const items = [];

                snapshot.docs.forEach((doc) => {
                    items.push({ ...doc.data(), id: doc.id });
                });
                setCartItems(items);
            });
            return () => {
                // Detach the listener when the component unmounts
                unsubscribe();
            };
        };

        fetchCartItems();
    }, [userId]);

    const value = {
        addItemToCart,
        cartItems,
        cartCount,
        removeItemFromCart,
        updateCartItem,
    };

    return (
        <>
            <CartContext.Provider value={value}>
                {children}
            </CartContext.Provider>
            <ToastContainer />
        </>
    );
};
