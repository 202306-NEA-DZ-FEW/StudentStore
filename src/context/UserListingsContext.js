import { createContext, useContext, useEffect, useState } from "react";
import { onSnapshot, query, where } from "firebase/firestore";
import { db } from "@/util/firebase";
import { useAuth } from "./AuthContext";

export const UserListingsContext = createContext({
    userSaleItemCount: 0,
    userBorrowItemCount: 0,
});

export const UserListingsProvider = ({ children }) => {
    const { currentUser } = useAuth();
    const currentUserUid = currentUser?.uid || "";
    const userId = currentUserUid;
    const [userSaleItemCount, setUserSaleItemCount] = useState(0);
    const [userBorrowItemCount, setUserBorrowItemCount] = useState(0);

    useEffect(() => {
        const fetchUserListings = async () => {
            const userSaleItemsQuery = query(
                db.collection("products"),
                where("currentUserUid", "==", userId),
                where("type", "==", "sale")
            );

            const userBorrowItemsQuery = query(
                db.collection("products"),
                where("currentUserUid", "==", userId),
                where("type", "==", "borrow")
            );

            const userSaleItemsUnsubscribe = onSnapshot(
                userSaleItemsQuery,
                (snapshot) => {
                    setUserSaleItemCount(snapshot.docs.length);
                }
            );

            const userBorrowItemsUnsubscribe = onSnapshot(
                userBorrowItemsQuery,
                (snapshot) => {
                    setUserBorrowItemCount(snapshot.docs.length);
                }
            );

            return () => {
                // Detach the listeners when the component unmounts
                userSaleItemsUnsubscribe();
                userBorrowItemsUnsubscribe();
            };
        };

        if (userId) {
            fetchUserListings();
        }
    }, [userId]);

    const value = {
        userSaleItemCount,
        userBorrowItemCount,
    };

    return (
        <UserListingsContext.Provider value={value}>
            {children}
        </UserListingsContext.Provider>
    );
};
