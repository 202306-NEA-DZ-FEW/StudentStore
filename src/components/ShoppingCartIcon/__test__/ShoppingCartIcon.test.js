import renderer from "react-test-renderer";

import { AuthProvider } from "@/context/AuthContext";

import ShoppingCartIcon from "../ShoppingCartIcon";

jest.mock("../../../util/firebase", () => {
    return {
        initializeApp: jest.fn(),
    };
});

it("renders correctly", () => {
    const tree = renderer
        .create(
            <AuthProvider>
                <ShoppingCartIcon />
            </AuthProvider>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
