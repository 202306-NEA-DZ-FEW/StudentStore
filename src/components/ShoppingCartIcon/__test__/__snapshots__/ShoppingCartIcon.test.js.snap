// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <a
    className="relative"
    href="/cart"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
  >
    <svg
      className="className=' text-[28px] text-[#8685a7] hover:text-orange-500 "
      fill="none"
      height="1em"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="9"
        cy="21"
        r="1"
      />
      <circle
        cx="20"
        cy="21"
        r="1"
      />
      <path
        d="M1 1h4l2.68 13.39a2 2 0 0 0 2 1.61h9.72a2 2 0 0 0 2-1.61L23 6H6"
      />
    </svg>
  </a>
</div>
`;
